### PR TITLE
 mgr/dashboard: RGW buckets async validator performance enhancement and name constraints

### PR DESCRIPTION
**Pull request by nizamial09**
The rgw bucket creation form has the Name field which have an async
validator. The validator calls all the bucket name and check if the
entered name is unique or not. This happens on every keystroke. So if
100 or more buckets are there, then the async validation can be real
slow and causes misvalidations in different fields.

I changed the validation logic and did some cleanups to improve the
performance of the async validation.

**BEFORE**

https://user-images.githubusercontent.com/71764184/116041569-fa4d4e00-a68a-11eb-99cd-55b1505da7ba.mp4

**AFTER**

https://user-images.githubusercontent.com/71764184/116041593-02a58900-a68b-11eb-983a-6419db0d61fa.mp4

Also explains the rgw bucket name constrains for each bucket name validation
errors.

Fixes: https://tracker.ceph.com/issues/50514
Fixes: https://tracker.ceph.com/issues/50516
Signed-off-by: Nizamudeen A <nia@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 1]
```diff
@@ -50,6 +50,18 @@
...
+                  i18n>Bucket names must not contain uppercase characters or underscores.</span>
+            <span class="invalid-feedback"
+                  *ngIf="bucketForm.showError('bid', frm, 'lowerCaseOrNumber')"
+                  i18n>Each label must start and end with a lowercase letter or a number.</span>
+            <span class="invalid-feedback"
+                  *ngIf="bucketForm.showError('bid', frm, 'ipAddress')"
+                  i18n>Bucket names cannot be formatted as IP address.</span>
+            <span class="invalid-feedback"
+                  *ngIf="bucketForm.showError('bid', frm, 'shouldBeInRange')"
+                  i18n>Bucket names can be between 3 and 63 characters long.</span>
```
Comment (aaSharma14) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:64]:
```suggestion
                  i18n>Bucket names must be 3 to 63 characters long.</span>
```

**Review by aaSharma14 — COMMENTED** [review]
Just a suggestion.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts [hunk 2]
```diff
@@ -239,59 +240,88 @@ export class RgwBucketFormComponent extends CdForm implements OnInit {
...
+      // or the control value is empty.
+      if (control.pristine || control.value === '') {
+        return observableOf(null);
+      }
+      const constraints = [];
+      let errorName: string;
+      // - Bucket names cannot be formatted as IP address.
+      constraints.push(() => {
+        const ipv4Rgx = /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/i;
+        const ipv6Rgx = /^(?:[a-f0-9]{1,4}:){7}[a-f0-9]{1,4}$/i;
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:254]: Aren't there other places in our code where IPv4/6 needs be validated so we make this reusable? Or couldn't we replace part of this validation logic with some external validator library (I did a quick check but [no luck](https://www.npmjs.com/package/ngx-custom-validators))?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html (hunk 1 above)
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:64]: Is there any way to reduce the amount of boilerplate code here? Perhas directly return the error message from the validator (or is that an antipatter... I'm not too familiar with this part of Angular)?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts (hunk 2 above)
Comment (pereman2) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:254]: I think we should add test for this
```js
const tests = [
'1.1.1.01',
'001.1.1.01',
]
const ipv4Rgx = /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/i;
for(let t in tests)
{
	const b = ipv4Rgx.test(tests[t])
	console.log(b)
}
```
output:
```
true
true
```
I don't think anyone will type that but who knows

Comment (pereman2) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:254]: I think we should add test for this
```js
const tests = [
'1.1.1.01',
'001.1.1.01',
]
const ipv4Rgx = /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/i;
for(let t in tests)
{
	const b = ipv4Rgx.test(tests[t])
	console.log(b)
}
```
output:
```
true
true
```
I don't think anyone will type that but who knows

Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:257]: @epuertat There's a `CdValidators.ip()` we are using, but I was not succesfull in making it work over here. It was actually used here previously but its actually incorrectly validating the fields (even if we enter IP address it doesn't give error). I checked the external libraries but none for the IP address.

**Comment by nizamial09** [comment]
jenkins retest this please

**Comment by nizamial09** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html (hunk 1 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html]: Nope, LGTM!

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts [hunk 3]
```diff
@@ -242,59 +243,88 @@ export class RgwBucketFormComponent extends CdForm implements OnInit {
...
+              case 'lowerCaseOrNumber':
+                return { lowerCaseOrNumber: true };
+              default:
+                return { bucketNameInvalid: true };
+            }
+          })
+        );
+      }
+      // - Bucket names must be unique.
+      return observableTimer().pipe(
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:318]: I think here also some dueTime should be added.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts [hunk 4]
```diff
@@ -54,94 +54,79 @@ describe('RgwBucketFormComponent', () => {
...
-            expect(resp).toBe(null);
-          } else {
-            expect(resp instanceof Object).toBeTruthy();
-            expect(resp.bucketNameInvalid).toBeTruthy();
-          }
-        });
+    const testValidator = (name: string, valid: boolean, expectedError?: string) => {
+      rgwBucketServiceGetSpy.and.returnValue(throwError('foo'));
+      formHelper.setValue('bid', name, true);
+      tick(500);
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts:60]:
```suggestion
      tick();
```
You removed the delay time of 500, right? So why `tick(500) ` here then?

**Comment by nizamial09** [comment]
@alfonsomthd @avanthakkar Took care of that weird behaviour.

**Comment by nizamial09** [comment]
> @alfonsomthd @avanthakkar Took care of that weird behaviour. and yeah dashes are allowed but not at the end

**Review by avanthakkar — APPROVED** [review]
Great work @nizamial09, LGTM!

**Comment by avanthakkar** [comment]
@nizamial09 make-check failure: 
```
[lint:tslint  ] Linting "ceph-dashboard"...
[lint:tslint  ] /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:294:42
[lint:tslint  ] ERROR: 294:42  whitespace  missing whitespace
[lint:tslint  ] ERROR: 294:43  whitespace  missing whitespace
```

**Comment by nizamial09** [comment]
@epuertat I created one tracker for it: https://tracker.ceph.com/issues/50762 I'll look for some alternates to return these errors.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts [hunk 5]
```diff
@@ -54,94 +54,87 @@ describe('RgwBucketFormComponent', () => {
...
-    });
+    it('should validate empty name', fakeAsync(() => {
+      formHelper.expectErrorChange('bid', '', 'required', true);
+    }));
 
-    it('bucket names cannot be formatted as IP address', () => {
-      testValidator('172.10.4.51', false);
-    });
+    it('bucket names cannot be formatted as IP address', fakeAsync(() => {
+      const testIPs = ['1.1.1.01', '001.1.1.01', '127.0.0.1'];
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts:73]: You can also add an IPv6 like 2001:0db8:85a3:0000:0000:8a2e:0370:7334

**Review by alfonsomthd — APPROVED** [review]

**Comment by epuertat** [comment]
> @epuertat I created one tracker for it: [tracker.ceph.com/issues/50762](https://tracker.ceph.com/issues/50762) I'll look for some alternates to return these errors.

Thanks @nizamial09! I liked what [I read in this link](https://medium.com/swlh/angular-validation-errors-made-easy-3bc1a837c6e7) that @avanthakkar shared.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts (hunk 5 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts:73]: Sorry @alfonsomthd I didn't had a chance to take a look at this in the morning.

**Pull request merged**